res_prometheus: Do not generate broken metrics

zecke opened the pull request:
In 8d6fdf9c3adede201f0ef026dab201b3a37b26b6 invisible bridges were skipped but that lead to producing metrics with no name and no help.

Keep track of the number of metrics configured and then only emit these. Add a basic testcase that verifies that there is no '(NULL)' in the output.

ASTERISK-30474

gtjoseph commented:
If this PR applies to other branches, please add a comment with the appropriate "cherry-pick-to" headers as per the [Create a Pull Request](https://wiki.asterisk.org/wiki/display/AST/Code+Contribution#CodeContribution-CreateaPullRequest) process.  If you don't want it cherry-picked, please add a comment stating "No cherry-picks required" so we don't keep asking.

zecke commented:
cherry-pick-to: 20
cherry-pick-to: 19
cherry-pick-to: 18

gtjoseph commented:
I removed 19 from the cherry-pick branches as that branch is in security-fix-only mode.

seanbright commented on tests/test_res_prometheus.c:
```diff
@@ -699,6 +702,52 @@ AST_TEST_DEFINE(config_general_core_metrics)
+	if (!bridge) {
+		return;
+	}
+	ast_bridge_destroy(bridge, 0);
```
The documentation for `ast_bridge_destroy(...)` indicates that that the bridge argument must be locked on entry.

zecke replied:
Wouldn't this lead to a deadlock: https://github.com/asterisk/asterisk/blob/869cb0c2608c7602eb372063a28f5160b57ac412/main/bridge.c#L947?

seanbright replied:
Apologies for this. I’m not sure what documentation I was looking at but clearly it was not applicable.

seanbright commented:
Also, can you update the commit message to reference 8d6fdf9c3adede201f0ef026dab201b3a37b26b6 rather than I58ef9f44036feded5966b5fc70ae754f8182883d since gerrit is no more?

zecke commented:
@seanbright could you please have another look. I updated the commit message but I am not sure whether locking the bridge is the right thing.

zecke commented:
> > @seanbright could you please have another look. I updated the commit message but I am not sure whether locking the bridge is the right thing.
> 
> The bridge should not be locked when ast_bridge_destroy is called.
> 
> Additionally you've now got a second merge commit, this should be squashed down to a single commit on this PR.

I ended up pressing the "Update branch" button in this UI. Rebased, forced push, reran the tests. Anything else that I should be doing?

jcolp commented:
Nope.

zecke commented:
> Nope.

I tried to re-read https://wiki.asterisk.org/wiki/display/AST/Code+Contribution but I am not sure what to expect in terms of getting this merged. Anything I am missing?

jcolp commented:
No, you're not missing anything. It requires someone else to do a review and then if approved would be merged in.

vinzens commented:
Is there anybody else able to approve this PR? Would be great if Prometheus becomes usable.

jcolp commented:
I've asked this to be looked at by someone else. Apparently it wasn't showing up for others.
